crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move

```diff
@@ -401,6 +432,143 @@ module sui::kiosk {
+    // - Following the security-first approach and making sure that the Kiosk
+    // Owner and their assets are protected from malicious activity.
+    // - Making sure that the Kiosk Owner can disable any extension at any time
+    // and that the extension cannot hide its data from the Owner.
+    //
+    // Extension states:
+    // - Enabled: the extension is enabled by the owner and can be used.
+    // - Disabled: the extension is disabled. No methods can be called on it.
+    // - Removed: the extension is completely removed from the Kiosk.
+
+    /// Add an extension to the `Kiosk`. Can only be performed by the Kiosk Owner,
+    /// the visibility is intentionally made `entry` so that the call cannot be
+    /// hidden and called arbitrarily.
+    entry fun add_extension<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+        permissions: u32,
+        ctx: &mut TxContext
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        assert!(permissions <= permissions::max_permissions(), EIncorrectPermissions);
+        df::add(&mut self.id, ExtensionKey<Ext> { enabled: true }, Extension {
+            storage: bag::new(ctx),
+            permissions,
+            types: vector[]
+        });
+    }
+
+    public fun revoke_permissions<Ext: drop>(self: &mut Kiosk, cap: &KioskOwnerCap) {
+
+    }
+
+    /// Re-enables a disabled extension of the `Kiosk`. Can only be performed by the Kiosk Owner.
+    public fun enable_extension<Ext: drop>(self: &mut Kiosk, cap: &KioskOwnerCap) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        assert!(df::exists_(&self.id, ExtensionKey<Ext> { enabled: false }), EExtensionNotFound);
+
+        let ext: Extension = df::remove(&mut self.id, ExtensionKey<Ext> { enabled: false });
+        df::add(&mut self.id, ExtensionKey<Ext> { enabled: true }, ext);
+    }
+
+    /// Disables an extension of the `Kiosk`. Can only be performed by the Kiosk Owner.
+    public fun disable_extension<Ext: drop>(self: &mut Kiosk, cap: &KioskOwnerCap) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        assert!(df::exists_(&self.id, ExtensionKey<Ext> { enabled: true }), EExtensionNotFound);
+
+        let ext: Extension = df::remove(&mut self.id, ExtensionKey<Ext> { enabled: true });
+        df::add(&mut self.id, ExtensionKey<Ext> { enabled: false }, ext);
+    }
+
+    /// Uninstalls a disabled extension of the `Kiosk`. Can only be performed by the Kiosk Owner.
+    public fun uninstall_extension<Ext: drop>(self: &mut Kiosk, cap: &KioskOwnerCap) {
+        let key = ExtensionKey<Ext> { enabled: false };
+
+        assert!(object::id(self) == cap.for, ENotOwner);
+        assert!(df::exists_(&self.id, key), EExtensionNotDisabled);
+
+        let Extension {
+            storage,
+            permissions: _,
+            types: _,
+        } = df::remove(&mut self.id, key);
+
+        assert!(bag::is_empty(&storage), EExtensionStorageNotEmpty);
```

Comment:
Won't the attempt to delete throw an `EBagNotEmpty` anyways? _(Trying to understand how we decide to throw our own errors vs allowing the core ones)_

Reply:
The preference is always to have errors local to the module. So you don't get "module: dynamic_field; code: 2" scenario.